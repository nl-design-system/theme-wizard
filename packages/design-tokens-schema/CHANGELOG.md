# @nl-design-system-community/design-tokens-schema

## 1.0.0

### Major Changes

- 396ca51: Initial release. Very much work in progress!
