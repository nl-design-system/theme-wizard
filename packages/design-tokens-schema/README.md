# Design Tokens Schema

This repository contains Zod schemas and Typescript types to validate Design Token configurations.

## Features

- 🛂 Inspect design token configuration to be valid
- 🎨 Accept legacy tokens as well as modern, spec-compliant tokens
- 🛠️ Upgrade legacy-format tokens to modern, spec-compliant tokens
- 🫵 Optionally resolve design token references
- 🤏 Fine-grained schemas and types for maximum composability

## Installation

```sh
npm install @nl-design-system-community/design-tokens-schema
```

## Usage

### Validating an entire theme configuration

```ts
import { ThemeSchema, type Theme resolveConfigRefs } from '@nl-design-system-community/design-token-schema';

const myConfig = {
  brand: {
    ma: {
      color: {
        black: { $value: '#000' } }
        // + dozens of tokens
      }
    }
  },
  common: {
    basis: {
      color: {
        default:
          'bg-document': {
            $type: 'color',
            $value: '{ma.color.black}'
          }
        }
      }
    }
  }
};

// All refs like `{ma.color.black}` refs are validated to exist in `brand`.
const config = ThemeSchema.parse(myConfig) satisfies Theme;

// All refs like `{ma.color.black}` are first validated in the schema and the `.transform(resolveConfigRefs)` replaces the refs with the actual values
const resolvedConfig = ThemeSchema.transform(resolveConfigRefs).parse(myConfig) satisfies Theme;
```
