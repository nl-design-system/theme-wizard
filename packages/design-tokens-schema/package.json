{
  "name": "@nl-design-system-community/design-tokens-schema",
  "version": "1.0.0",
  "author": "Community for NL Design System",
  "description": "Zod schema's for Design Tokens JSON format",
  "license": "EUPL-1.2",
  "keywords": [
    "nl-design-system"
  ],
  "type": "module",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs"
    }
  },
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "access": "public",
    "provenance": true
  },
  "repository": {
    "type": "git+ssh",
    "url": "git@github.com:nl-design-system/theme-wizard.git",
    "directory": "packages/design-tokens-schema"
  },
  "scripts": {
    "test": "vitest",
    "test-coverage": "vitest run --coverage",
    "build": "vite build",
    "clean": "rimraf dist/"
  },
  "devDependencies": {
    "@vitest/coverage-v8": "3.2.4",
    "rimraf": "6.0.1",
    "typescript": "5.9.3",
    "vite": "6.3.5",
    "vite-plugin-dts": "4.5.4",
    "vitest": "3.2.4"
  },
  "dependencies": {
    "colorjs.io": "0.5.2",
    "zod": "4.1.12"
  }
}
