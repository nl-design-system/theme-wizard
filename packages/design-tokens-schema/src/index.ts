export * from './base-token';
export * from './color-token';
export * from './dimension-token';
export * from './fontfamily-token';
export * from './basis-tokens';
