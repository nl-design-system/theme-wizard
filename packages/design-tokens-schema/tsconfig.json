{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "lib": ["es2022", "dom"],
    "outDir": "dist",
    "skipLibCheck": true,
    "moduleResolution": "bundler"
  },
  "include": ["src"],
  "exclude": ["**/*.test.ts"]
}
