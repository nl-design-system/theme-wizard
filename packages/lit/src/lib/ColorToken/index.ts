import type { ScrapedColorToken } from '@nl-design-system-community/css-scraper';
import { ColorSpace, ColorComponent, COLOR_SPACES } from '@nl-design-system-community/design-tokens-schema';
export type { ColorSpace };

export type ColorTokenOkLCh = ColorToken & {
  $value: ColorToken['$value'] & {
    colorSpace: 'oklch';
  };
};

export type ColorComponents = ScrapedColorToken['$value']['components'];

type MinimalScrapedColorToken = Omit<ScrapedColorToken, '$extensions'> & {
  $extensions?: Partial<ScrapedColorToken['$extensions']>;
};

export default class ColorToken {
  #$type = 'color' as const;
  #$value: ScrapedColorToken['$value'];
  #$extensions: MinimalScrapedColorToken['$extensions'];

  constructor(token: MinimalScrapedColorToken) {
    const { $extensions, $value } = token;
    this.#$extensions = $extensions || {};
    this.#$value = $value;
  }

  get $type() {
    return this.#$type;
  }

  get $value() {
    return this.#$value;
  }

  get $extensions() {
    return this.#$extensions;
  }

  toCSSColorFunction(): string {
    return ColorToken.getCSSColorFunction(this.$value);
  }

  static getCSSColorFunction({
    colorSpace,
    components,
  }: {
    colorSpace: ColorSpace;
    components: ColorComponents;
  }): string {
    /**
     * @SEE https://www.designtokens.org/tr/third-editors-draft/color/#supported-color-spaces
     */
    const percentage = (c: ColorComponent) => (c === 'none' ? c : `${c}%`);

    const [first, ...rest] = components;

    switch (colorSpace) {
      // These color spaces are polar:
      case COLOR_SPACES.HSL:
        return `hsl(${first} ${rest.map(percentage).join(' ')})`;
      case COLOR_SPACES.HWB:
        return `hwb(${first} ${rest.map(percentage).join(' ')})`;
      case COLOR_SPACES.LCH:
        return `lch(${percentage(first)} ${rest.join(' ')})`;
      case COLOR_SPACES.OKLCH:
        return `oklch(${components.join(' ')})`;
      // These color spaces are cartesian:
      case COLOR_SPACES.LAB:
        return `lab(${percentage(first)} ${rest.join(' ')})`;
      case COLOR_SPACES.OKLAB:
        return `oklab(${components.join(' ')})`;
      // These color spaces are cartesian and have components encoded as [ 0.0 - 1.0, 0.0 - 1.0 , 0.0 - 1.0 ]
      case COLOR_SPACES.SRGB:
      case COLOR_SPACES.DISPLAY_P3:
      case COLOR_SPACES.A98_RGB:
      case COLOR_SPACES.PROPHOTO_RGB:
      case COLOR_SPACES.REC2020:
      case COLOR_SPACES.XYZ_D50:
      case COLOR_SPACES.XYZ_D65:
      default:
        // Assume a missing color space is also cartesian
        return `color(${colorSpace} ${components.join(' ')})`;
    }
  }
}
