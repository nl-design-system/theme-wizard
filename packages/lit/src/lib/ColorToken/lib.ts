import { COLOR_SPACES, ColorSpace } from '@nl-design-system-community/design-tokens-schema';

export function createHelperElement() {
  const canRunInBrowser = CSS?.supports('color', 'oklch(from red l c h');
  if (!canRunInBrowser) return null;

  const element = document.createElement('div');
  element.style.display = 'none';
  document.body.appendChild(element);
  return element;
}

const numberPattern = '(-?[\\d.]+(?:e[+-]?\\d+)?)';
const colorComponentRegex = new RegExp(`${numberPattern}\\s+${numberPattern}\\s+${numberPattern}\\)$`);

export function getCSSColorComponents(value: string) {
  const [, a, b, c] = colorComponentRegex.exec(value)?.map(Number) || [NaN, NaN, NaN, NaN];
  const components: [number, number, number] = [a, b, c];
  return components;
}

export function limitColorComponents(
  colorSpace: ColorSpace,
  components: [number, number, number],
): [number, number, number] {
  const percentage = (value: number) => Math.max(0, Math.min(value, 100));
  const positive = (value: number) => Math.max(0, value);
  const fraction = (value: number) => Math.max(0, Math.min(value, 1));

  const [a, b, c] = components.map(Number);

  switch (colorSpace) {
    case COLOR_SPACES.HSL:
      return [a, percentage(b), percentage(c)];
    case COLOR_SPACES.HWB:
      return [a, percentage(b), percentage(c)];
    case COLOR_SPACES.LAB:
      return [percentage(a), b, c];
    case COLOR_SPACES.LCH:
      return [percentage(a), positive(b), c];
    case COLOR_SPACES.OKLAB:
      return [fraction(a), b, c];
    case COLOR_SPACES.OKLCH:
      return [fraction(a), positive(b), c];
    case COLOR_SPACES.SRGB:
    case COLOR_SPACES.DISPLAY_P3:
    case COLOR_SPACES.A98_RGB:
    case COLOR_SPACES.PROPHOTO_RGB:
    case COLOR_SPACES.REC2020:
    case COLOR_SPACES.XYZ_D50:
    case COLOR_SPACES.XYZ_D65:
    default:
      return [fraction(a), fraction(b), fraction(c)];
  }
}

type Degree = number;
/**
 * Get hue component for a specific RGB value.
 *
 * @SEE https://en.wikipedia.org/wiki/Hue
 *
 * @param RGB as numbers between 0 and 1, inclusive
 * @returns number between 0 - 360, latter exclusive
 */
export function getHue([r, g, b]: [r: number, g: number, b: number]): Degree {
  const radians = Math.atan2(Math.sqrt(3) * (g - b), 2 * r - g - b);
  return ((radians * 180) / Math.PI + 360) % 360; // wrap around when negative
}

type Percentage = number;
/**
 * Convert RGB components to HSL components.
 *
 * @SEE https://en.wikipedia.org/wiki/HSL_and_HSV
 *
 * @param RGB as numbers between 0 and 1, inclusive
 * @returns Hue in degrees, saturation and lightness in percentages
 */
export function toHSL([r, g, b]: [r: number, g: number, b: number]): [Degree, Percentage, Percentage] {
  const hue = getHue([r, g, b]);
  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);
  const lightness = (max + min) / 2;
  const saturation =
    lightness % 1 !== 0 // no divide by zero
      ? (max - lightness) / Math.min(lightness, 1 - lightness)
      : 0;
  return [hue, saturation * 100, lightness * 100];
}

/**
 * Convert RGB components to HWB components.
 *
 * @SEE https://en.wikipedia.org/wiki/HWB_color_model
 *
 * @param RGB as numbers between 0 and 1, inclusive
 * @returns Hue in degrees, whiteness and blackness in percentages
 */
export function toHWB([r, g, b]: [r: number, g: number, b: number]): [Degree, Percentage, Percentage] {
  const hue = getHue([r, g, b]);
  const whiteness = Math.min(r, g, b);
  const blackness = 1 - Math.max(r, g, b);
  return [hue, whiteness * 100, blackness * 100];
}
