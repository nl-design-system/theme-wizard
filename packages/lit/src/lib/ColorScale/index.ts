import ColorToken from '../ColorToken';

export default class ColorScale {
  constructor(token: ColorToken) {}
}
