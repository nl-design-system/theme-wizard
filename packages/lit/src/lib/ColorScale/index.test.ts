import { COLOR_SPACES } from '@nl-design-system-community/design-tokens-schema';
import { describe, expect, test } from 'vitest';
import ColorToken from '../ColorToken';
import ColorScale from './index';

const greenSRGB = new ColorToken({
  $value: {
    alpha: 1,
    colorSpace: COLOR_SPACES.SRGB,
    components: [0, 0.5058823529411764, 0.12156862745098039],
  },
});

const orangeSRGB = new ColorToken({
  $value: {
    alpha: 1,
    colorSpace: COLOR_SPACES.SRGB,
    components: [0.9372549019607843, 0.49019607843137253, 0],
  },
});

const greenOKLCH = new ColorToken({
  $value: {
    ...greenSRGB.$value,
    colorSpace: COLOR_SPACES.OKLCH,
    components: [0.524144, 0.165652, 144.827],
  },
});

const orangeOKLCH = new ColorToken({
  $value: {
    ...orangeSRGB.$value,
    colorSpace: COLOR_SPACES.OKLCH,
    components: [0.705249, 0.17366, 55.4715],
  },
});

describe('ColorScale', () => {
  test('can be initialized with a token', () => {
    const colorScale = new ColorScale(greenOKLCH);
    expect(colorScale.from.toObject()).toMatchObject(greenOKLCH.toObject());
  });

  test('converts srgb to oklch on initialization', () => {
    const colorScale = new ColorScale(greenSRGB);
    expect(colorScale.fromOKLCH.toObject()).toMatchObject(greenOKLCH.toObject());
  });

  test('can be updated after initialization', () => {
    const colorScale = new ColorScale(greenSRGB);
    colorScale.from = orangeOKLCH;
    expect(colorScale.from.toObject()).toMatchObject(orangeOKLCH.toObject());
  });

  test('converts after update', () => {
    const colorScale = new ColorScale(greenSRGB);
    colorScale.from = orangeSRGB;
    expect(colorScale.fromOKLCH.toObject()).toMatchObject(orangeOKLCH.toObject());
  });

  test('returns a list of derived colors', () => {
    const colorScale = new ColorScale(greenSRGB);
    expect(colorScale.list().length).toBe(colorScale.size);
  });

  test('get token in color scale by position', () => {
    const colorScale = new ColorScale(greenSRGB);
    for (let index = 0; index++; index < colorScale.size) {
      expect(colorScale.get(index)).toBeDefined();
    }
  });

  test('list of derived colors contains base color exactly once, when lightness is in the mask', () => {
    const lightnessMatchedGreenOKLCH = new ColorToken({
      $value: {
        ...greenOKLCH.$value,
        components: [0.52, greenOKLCH.$value.components[1], greenOKLCH.$value.components[2]],
      },
    });
    const colorScale = new ColorScale(lightnessMatchedGreenOKLCH);
    const matches = colorScale
      .list()
      .filter(
        (color) => color.$value.components.toString() === lightnessMatchedGreenOKLCH.$value.components.toString(),
      );
    expect(matches.length).toBe(1);
  });

  test('list of derived colors contains base color exactly once, even if lightness is not in the mask', () => {
    const colorScale = new ColorScale(greenSRGB);
    const matches = colorScale
      .list()
      .filter((color) => color.$value.components.toString() === greenSRGB.$value.components.toString());
    expect(matches.length).toBe(1);
  });

  test('exports scale to an object defining design tokens', () => {
    const colorScale = new ColorScale(greenSRGB);
    expect(colorScale.toObject()).toMatchSnapshot();
  });
});
