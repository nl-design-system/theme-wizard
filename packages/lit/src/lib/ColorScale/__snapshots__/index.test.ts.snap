// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorScale > exports scale to an object defining design tokens 1`] = `
{
  "1": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.967155,
        1.01514,
        0.966316,
      ],
    },
  },
  "10": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.146516,
        0.337337,
        0.087101,
      ],
    },
  },
  "11": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.145318,
        0.269897,
        0.0959797,
      ],
    },
  },
  "12": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.130491,
        0.204147,
        0.0860233,
      ],
    },
  },
  "13": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.0682464,
        0.0779915,
        0.0291205,
      ],
    },
  },
  "14": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.0383531,
        0.0378082,
        0.0115605,
      ],
    },
  },
  "2": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.896923,
        0.992622,
        0.895615,
      ],
    },
  },
  "3": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.837593,
        0.981392,
        0.836472,
      ],
    },
  },
  "4": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.777058,
        0.969582,
        0.776913,
      ],
    },
  },
  "5": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.714924,
        0.957199,
        0.716743,
      ],
    },
  },
  "6": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.525795,
        0.814327,
        0.532932,
      ],
    },
  },
  "7": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.1895,
        0.539467,
        0.221727,
      ],
    },
  },
  "8": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0,
        0.5058823529411764,
        0.12156862745098039,
      ],
    },
  },
  "9": {
    "$type": "color",
    "$value": {
      "alpha": 1,
      "colorSpace": "srgb",
      "components": [
        0.0958433,
        0.453323,
        0.0449238,
      ],
    },
  },
}
`;
